### PR TITLE
fwbuilder: add wayland support

### PR DESCRIPTION
**Pull request by elatov**
###### Description of changes
- adding wayland support (https://github.com/NixOS/nixpkgs/issues/171845)

###### Things done

- Built on platform(s)
  - [ x ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/security/fwbuilder/default.nix [hunk 1]
```diff
@@ -14,7 +25,8 @@ stdenv.mkDerivation rec {
   nativeBuildInputs = [
     cmake
     wrapQtAppsHook
-  ];
+  ]
+  ++ optionals waylandSupport [ wayland wayland-protocols qtwayland ];
```
Comment (Artturin) [inline pkgs/tools/security/fwbuilder/default.nix:29]: they should be in buildInputs

File: pkgs/tools/security/fwbuilder/default.nix [hunk 2]
```diff
@@ -1,4 +1,15 @@
-{ stdenv, lib, fetchFromGitHub, cmake, qtbase, wrapQtAppsHook }:
+{ lib
+, stdenv
+, fetchFromGitHub
+, cmake
+, qtbase
+, wrapQtAppsHook
+, waylandSupport ? true, wayland, wayland-protocols, qtwayland
```
Comment (Artturin) [inline pkgs/tools/security/fwbuilder/default.nix:7]: doesn't have to be an option

no need to specify qtwayland

File: pkgs/tools/security/fwbuilder/default.nix [hunk 3]
```diff
@@ -16,6 +30,10 @@ stdenv.mkDerivation rec {
     wrapQtAppsHook
   ];
 
+  buildInputs = [
```
Comment (Artturin) [inline pkgs/tools/security/fwbuilder/default.nix:33]: No need to have []++

File: pkgs/tools/security/fwbuilder/default.nix (hunk 2 above)
Comment (Artturin) [reply pkgs/tools/security/fwbuilder/default.nix]: .

**Comment by Artturin** [comment]
i just tested and the wrapper already has qtwayland in it so you don't need to add that

**Comment by elatov** [comment]
@anthr76 would you mind letting us know how you are installing/setting up wayland. I ended up creating this config:

```
$ cat /etc/nixos/way.nix 
{ config, pkgs, ... }:

{
  services.xserver = {
    enable = true;
    displayManager.gdm.enable = true;
    displayManager.gdm.wayland = true;
    desktopManager.gnome.enable = true;
  };
  programs.sway.enable = true;
  environment.systemPackages = with pkgs; [
    git
  ];
}
```

Then running:

```
sudo nixos-rebuild switch --upgrade
```

And then following these command to install the new version:

```
export NIXPKGS=/data/nixpkgs
nix-build $NIXPKGS -A fwbuilder
```

And it works for me, even the above flags.

File: pkgs/tools/security/fwbuilder/default.nix [hunk 4]
```diff
@@ -1,4 +1,18 @@
-{ stdenv, lib, fetchFromGitHub, cmake, qtbase, wrapQtAppsHook }:
+{ lib
+, stdenv
+, fetchFromGitHub
+, cmake
+, qtbase
+, wrapQtAppsHook
+, waylandSupport ? true
```
Comment (SuperSandro2000) [inline pkgs/tools/security/fwbuilder/default.nix:7]:
```suggestion
```

File: pkgs/tools/security/fwbuilder/default.nix [hunk 5]
```diff
@@ -1,4 +1,18 @@
...
+, wayland
+, wayland-protocols
+, qtwayland
+}:
+
+let
+  inherit (lib) optionalString optional optionals;
+in
```
Comment (SuperSandro2000) [inline pkgs/tools/security/fwbuilder/default.nix:16]:
```suggestion
```

File: pkgs/tools/security/fwbuilder/default.nix [hunk 6]
```diff
@@ -16,6 +30,10 @@ stdenv.mkDerivation rec {
     wrapQtAppsHook
   ];
 
+  buildInputs = optionals waylandSupport [
```
Comment (SuperSandro2000) [inline pkgs/tools/security/fwbuilder/default.nix:33]:
```suggestion
  buildInputs = lib.optionals waylandSupport [
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/tools/security/fwbuilder/default.nix (hunk 4 above)
Comment (elatov) [reply pkgs/tools/security/fwbuilder/default.nix:7]: removing this:

```
$ head ./pkgs/tools/security/fwbuilder/default.nix
{ lib
, stdenv
, fetchFromGitHub
, cmake
, qtbase
, wrapQtAppsHook
, wayland
, wayland-protocols
, qtwayland
}:
```

Makes the build fail:

```
$ nix-build $NIXPKGS -A fwbuilder
error: undefined variable 'waylandSupport' at /data/nixpkgs/pkgs/tools/security/fwbuilder/default.nix:28:31
```

And removing the optional flag also fails:

```
$ head ./pkgs/tools/security/fwbuilder/default.nix
{ lib
, stdenv
, fetchFromGitHub
, cmake
, qtbase
, wrapQtAppsHook
, waylandSupport
, wayland
, wayland-protocols
, qtwayland

[elatov@nixos:/data/nixpkgs]$ nix-build $NIXPKGS -A fwbuilder
error: Function called without required argument "waylandSupport" at /data/nixpkgs/pkgs/tools/security/fwbuilder/default.nix
```

Comment (Artturin) [reply pkgs/tools/security/fwbuilder/default.nix:7]: you, of course, need to remove the uses of it

**Comment by elatov** [comment]
@Artturin and @SuperSandro2000 did I miss anything in my last push?

Thanks,

File: pkgs/tools/security/fwbuilder/default.nix [hunk 7]
```diff
@@ -1,4 +1,13 @@
-{ stdenv, lib, fetchFromGitHub, cmake, qtbase, wrapQtAppsHook }:
+{ lib
+, stdenv
+, fetchFromGitHub
+, cmake
+, qtbase
+, wrapQtAppsHook
+, wayland
+, wayland-protocols
+, qtwayland
```
Comment (Artturin) [inline pkgs/tools/security/fwbuilder/default.nix:9]:
```suggestion
, qtwayland
```

Comment (SuperSandro2000) [reply pkgs/tools/security/fwbuilder/default.nix:9]: Thats used in line 31

Comment (elatov) [reply pkgs/tools/security/fwbuilder/default.nix:9]: @Artturin after removing that line the build fails:

```
$ head ./pkgs/tools/security/fwbuilder/default.nix
{ lib
, stdenv
, fetchFromGitHub
, cmake
, qtbase
, wrapQtAppsHook
, wayland
, wayland-protocols
}:
```

With this error:

```
$ nix-build $NIXPKGS -A fwbuilder
error: undefined variable 'qtwayland' at /data/nixpkgs/pkgs/tools/security/fwbuilder/default.nix:30:5
```

**Pull request merged**